crates/scheduler/src/schedule/mod.rs

```diff
@@ -0,0 +1,910 @@
+    /// Expected the `SystemExecutionGuard` channel to be dropped to signal completion, but got message.
+    #[error("expected the `SystemExecutionGuard` channel to be dropped to signal completion, but got message {0:?}")]
+    IncorrectSystemCompletionMessage(Result<(), RecvError>),
+    /// The schedule has deadlocked due to there being no pending systems but all systems have still not run.
+    #[error("the schedule has deadlocked due to there being no pending systems but all systems have still not run")]
+    Deadlock,
+}
+
+/// Whether a precedence graph operation succeeded.
+type PrecedenceGraphResult<T, E = PrecedenceGraphError> = Result<T, E>;
+
+/// Orders [`System`]s based on their precedence.
+#[derive(Default, Clone)]
+pub struct PrecedenceGraph<'systems> {
+    dag: SysDag<'systems>,
+    /// Systems which have been given out, and are awaiting execution.
+    ///
+    /// Warning: Node indices are _not_ stable and will be invalidated if [`dag`] is mutated.
+    pending: Vec<(Receiver<()>, NodeIndex)>,
+    /// Which systems have already been executed this tick.
+    ///
+    /// Warning: Node indices are _not_ stable and will be invalidated if [`dag`] is mutated.
+    already_executed: Vec<NodeIndex>,
+}
+
+impl<'systems> Debug for PrecedenceGraph<'systems> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        // Instead of printing the struct, format the DAG in dot-format and print that.
+        // (so it can be viewed through tools like http://viz-js.com/)
+        writeln!(f, "{:?}", daggy::petgraph::dot::Dot::new(self.dag.graph()))
+    }
+}
+
+impl<'systems> Display for PrecedenceGraph<'systems> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        // Instead of printing the struct, format the DAG in dot-format and print that.
+        // (so it can be viewed through tools like http://viz-js.com/)
+        writeln!(f, "{}", daggy::petgraph::dot::Dot::new(self.dag.graph()))
+    }
+}
+
+// PrecedenceGraph-equality should be based on whether the underlying DAGs are
+// isomorphic or not - i.e. whether they're equivalently constructed but not necessarily
+// exactly the same.
+impl<'systems> PartialEq<Self> for PrecedenceGraph<'systems> {
+    fn eq(&self, other: &Self) -> bool {
+        let node_match = |a: &Sys<'systems>, b: &Sys<'systems>| a == b;
+        let edge_match = |a: &i32, b: &i32| a == b;
+        daggy::petgraph::algo::is_isomorphic_matching(
+            &self.dag.graph(),
+            &other.dag.graph(),
+            node_match,
+            edge_match,
+        )
+    }
+}
+
+impl<'systems> Schedule<'systems> for PrecedenceGraph<'systems> {
+    fn generate(systems: &'systems [Box<dyn System>]) -> ScheduleResult<Self> {
+        let mut dag = Dag::new();
+
+        for system in systems.iter().map(|system| system.as_ref()) {
+            let node = dag.add_node(system);
+
+            // Find nodes which current system must run _after_ (i.e. current has dependency on).
```

Comment:
Why use markdown in normal comments when it is only supported in doc comments?

Reply:
To place extra emphasis on specific words, to make it clear for the reader that I want to stress a point in this comment.